plugins/redfish/fu-redfish-smc-device.c

```diff
@@ -0,0 +1,217 @@
+		return FALSE;
+	}
+	return fu_redfish_device_poll_task(FU_REDFISH_DEVICE(device), location, progress, error);
+}
+
+static gboolean
+fu_redfish_smc_device_write_firmware(FuDevice *device,
+				     FuFirmware *firmware,
+				     FuProgress *progress,
+				     FwupdInstallFlags flags,
+				     GError **error)
+{
+	FuRedfishSmcDevice *self = FU_REDFISH_SMC_DEVICE(device);
+	FuRedfishBackend *backend = fu_redfish_device_get_backend(FU_REDFISH_DEVICE(self));
+	CURL *curl;
+	JsonObject *json_obj;
+	curl_mimepart *part;
+	const gchar *location = NULL;
+	g_autoptr(curl_mime) mime = NULL;
+	g_autoptr(FuRedfishRequest) request = NULL;
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(GString) params = NULL;
+
+	/* get default image */
+	fw = fu_firmware_get_bytes(firmware, error);
+	if (fw == NULL)
+		return FALSE;
+
+	/* create the multipart for uploading the image request */
+	request = fu_redfish_backend_request_new(backend);
+	curl = fu_redfish_request_get_curl(request);
+	mime = curl_mime_init(curl);
+	(void)curl_easy_setopt(curl, CURLOPT_MIMEPOST, mime);
+
+	params = fu_redfish_smc_device_get_parameters(self);
+	part = curl_mime_addpart(mime);
+	curl_mime_name(part, "UpdateParameters");
+	(void)curl_mime_type(part, "application/json");
+	(void)curl_mime_data(part, params->str, CURL_ZERO_TERMINATED);
+	if (g_getenv("FWUPD_REDFISH_VERBOSE") != NULL)
+		g_debug("request: %s", params->str);
+
+	part = curl_mime_addpart(mime);
+	curl_mime_name(part, "UpdateFile");
+	(void)curl_mime_type(part, "application/octet-stream");
+	(void)curl_mime_filedata(part, "firmware.bin");
+	(void)curl_mime_data(part, g_bytes_get_data(fw, NULL), g_bytes_get_size(fw));
+
+	fu_progress_set_status(progress, FWUPD_STATUS_DEVICE_WRITE);
+	if (!fu_redfish_request_perform(request,
+					fu_redfish_backend_get_push_uri_path(backend),
+					FU_REDFISH_REQUEST_PERFORM_FLAG_LOAD_JSON,
+					error))
+		return FALSE;
+	if (fu_redfish_request_get_status_code(request) != 202) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INVALID_FILE,
+			    "failed to upload: %li",
+			    fu_redfish_request_get_status_code(request));
+		return FALSE;
+	}
+	json_obj = fu_redfish_request_get_json_object(request);
+
+	/* poll the verify task for progress */
+	location = fu_redfish_smc_device_get_task(json_obj);
+	if (location == NULL) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INVALID_FILE,
+			    "no task returned for %s",
+			    fu_redfish_backend_get_push_uri_path(backend));
+		return FALSE;
+	}
+	if (!fu_redfish_device_poll_task(FU_REDFISH_DEVICE(self), location, progress, error))
+		return FALSE;
+
+	return fu_redfish_smc_device_start_update(device, progress, error);
+}
+
+static void
+fu_redfish_smc_device_set_progress(FuDevice *self, FuProgress *progress)
+{
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 50, "write");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 50, "restart");
```

Comment:
I think you need 4 things here, no? don't you get a runtime warning?

Reply:
Before I got something like "error: progress reset to 0". Does the FuDevice code set the status or is it just my code?

Reply:
I think the docs are https://fwupd.github.io/libfwupdplugin/tutorial.html#set_progress -- but the redfish plugin is somewhat special as it wants to set the progress and status on it's own using the polling interface. Can you share a complete daemon log?

Reply:
[log.txt](https://github.com/fwupd/fwupd/files/9234178/log.txt)

Reply:
Another drive-by comment:

>   Guid:                 cea310e9-acad-5573-98c3-3f0322bc1886 <- REDFISH\VENDOR_SMCI&ID_BIOS

That's super generic. Does SuperMicro have any kind of ID that identifies the firmware *stream* -- i.e. is specific to this model and device?

Reply:
Can you do the same log again please, but this time with `FWUPD_REDFISH_VERBOSE=1` set pls.

Reply:
AFAIK all updates are mainboard specific and they don't have different stream/trains/LTS versions. This all the BMC gives me about the current version:
```
{
  "@odata.type": "#SoftwareInventory.v1_3_0.SoftwareInventory",
  "@odata.id": "/redfish/v1/UpdateService/FirmwareInventory/BIOS",
  "Id": "BIOS",
  "Name": "BIOS",
  "Updateable": true,
  "Version": "BIOS Date: 04/14/2022 Ver 2.4",
  "SoftwareId": "BIOS Date: 04/14/2022 Ver 2.4",
  "Status": {
    "State": "Enabled",
    "Health": "OK"
  },
  "RelatedItem": [
    {
      "@odata.id": "/redfish/v1/Systems/1/Bios"
    }
  ],
  "Actions": {
    "Oem": {}
  },
  "Oem": {},
  "Manufacturer": "SMCI",
  "ReleaseDate": "2022-04-14T00:00:00Z",
  "WriteProtected": false
}
```
The `/redfish/v1/Systems/1/Bios` route gives me a license error.

Reply:
Ha,

    "Version": "BIOS Date: 04/14/2022 Ver 2.4",
    "SoftwareId": "BIOS Date: 04/14/2022 Ver 2.4",

the latter isn't a software ID. I wonder if they copy&pasted something wrong here :) 

I guess we should use some SMBIOS data; what in `sudo fwupdtool hwids` makes sense? Perhaps `REDFISH\VENDOR_SMCI&ID_$ProductName$`?

Reply:
So maybe it isn't a huge problem if we 100% use a hwid to handle this. We can fix async to this feature, other Redfish stuff is going to be affected.

Reply:
tl;dr: not your problem

Reply:
[log-verbose-redfish.txt](https://github.com/fwupd/fwupd/files/9235858/log-verbose-redfish.txt)

Reply:
Ah, I see. There are two tasks that are polled. First the upload, then the actual update (`fu_redfisk_smc_device_start_update()`) maybe create a child progress for both?

Reply:
Yes, if you're the one that's doing the actions. Thanks!